### DISCUSSION TITLE
Question about web learning support

### DISCUSSION THREAD

**Discussion by ferganyalim** [LearnLib/learnlib#121] [Q&A]
Hi

I intend to add a command line tool for web learning to RAlib. From the release paper I read LearnLib supposedly has web learning support but I was not able to find it in the codebase. I wished to ask you if there are any already existing support for web learning whether in RAlib or LearnLib because otherwise I will implement it myself.

Thank you

**Comment by mtf90** [comment]
Dear @ferganyalim, support for learning web applications is implemented in [ALEX](https://github.com/LearnLib/alex) which is a standalone web application itself and allows you to conveniently (and collaboratively) set up learning projects for web applications. Internally, the tool uses [Selenium](https://www.selenium.dev/) to construct input symbols that are executed on a given web application.

Combining register automata with web application learning is an interesting idea. Unfortunately, ALEX is currently very self-contained, so there is no elegant way to reuse the Selenium definitions of ALEX in a library-like way. We have thought about this previously, but it would require a lot of refactoring in ALEX (not saying it is impossible, it just takes a lot of effort).

As for RAlib: While LearnLib/automatalib#5 and https://github.com/LearnLib/learnlib/issues/3 are already several years old, there are currently plans materializing to re-integrate RAlib into the main libraries in the upcoming months. Just as a heads-up so that you don't find yourself building against a moving target.

**Reply by ferganyalim** [reply]
Hello

Thank you for your reply. Yes I have originally considered extending ALEX but the internal code seemed to be very Mealy Machine spesific so I changed my focus to RALib instead. From your reply should I assume that current LearnLib has no support for web application SUL's? If that is so would you say that it would be reasonable for a single developer to develop in a timeframe of a month? Also I am producing this project as part of my dissertation hence I am looking towards a barebones implementation that gets the job done (hence a command line tool that uses a basic config file similar to class analyser tool of RALib) so I am not worried about RALib getting integrated in the future.

**Reply by mtf90** [reply]
LearnLib doesn't come with web-specific actions/SULs itself, but ALEX does hook into LearnLib interfaces to execute the learning algorithms. So you can just take the SUL/symbol/etc implementations of ALEX as a foundation and adjust them to your needs (it's open source after all).

You are correct in that ALEX focuses on Mealy machines only for now. Combining both approaches (e.g., by extending ALEX with register automata) would be a little bit cumbersome because both tools currently use different LearnLib versions. The versions are not conceptually different but there has been some incompatible refactorings between them. Especially given your time schedule, picking whichever components you need from the different projects and plugging them together in a custom CLI tool sounds like the most feasible approach for now.
